river_dl/train.py

```diff
@@ -21,33 +21,102 @@ def get_data_if_file(d):
+        self.model = model
+        self.model.compile(optimizer=optimizer, loss=loss_fn)
+        if weights:
+            self.model.load_weights(weights)
+
+    def pre_train(self, x, y, epochs, batch_size, out_dir):
+
+        ## Set up training log callback
+        csv_log = tf.keras.callbacks.CSVLogger(
+            os.path.join(out_dir, f"pretrain_log.csv")
+        )
+
+        # Use generic fit statement
+        self.model.fit(
+            x=x,
+            y=y,
+            epochs=epochs,
+            batch_size=batch_size,
+            callbacks=[csv_log],
+        )
+        # Save the pretrained weights
+        self.model.save_weights(os.path.join(out_dir, "pretrained_weights/"))
+        return self.model
+
+    def fine_tune(self, x, y, x_val, y_val, epochs, batch_size, out_dir, early_stop_rounds=False):
+        # Specify our training log
+        csv_log = tf.keras.callbacks.CSVLogger(
+            os.path.join(out_dir, "finetune_log.csv")
+        )
+
+        # Set up early stopping rounds if desired, setting this to number of epochs functionally doesn't use
+        # early stopping
+        if ~early_stop_rounds:
+            early_stop_rounds = epochs
+
+        early_stop = tf.keras.callbacks.EarlyStopping(
+            monitor='val_loss', min_delta=0, patience=early_stop_rounds, restore_best_weights=False,
+            verbose=1)
+
+        # Save alternate weight file that saves the best validation weights
+        best_val = tf.keras.callbacks.ModelCheckpoint(
+            os.path.join(out_dir, 'best_val_weights/'), monitor='val_loss', verbose=0, save_best_only=True,
+            save_weights_only=True, mode='min', save_freq='epoch')
+
+        # Ensure that training happens on CPU if using the GW loss function
+        if y.shape[2] > 2:
+            with tf.device('/CPU:0'):
+                self.model.fit(
+                    x=x,
+                    y=y,
+                    epochs=epochs,
+                    batch_size=batch_size,
+                    callbacks=[csv_log, early_stop, best_val],
+                    validation_data=(x_val, y_val)
+                )
+        else:
+           self.model.fit(
+               x=x,
+               y=y,
+               epochs=epochs,
+               batch_size=batch_size,
+               callbacks=[csv_log, early_stop, best_val],
+               validation_data=(x_val, y_val)
+        )
+
+        # Save our trained weights
+        self.model.save_weights(os.path.join(out_dir, f"trained_weights/"))
+        return self.model
+
+
 def train_model(
     io_data,
-    pretrain_epochs,
-    finetune_epochs,
+    epochs,
     hidden_units,
-    loss_func_ft,
+    loss_func,
     out_dir,
-    loss_func_pre = None,
     model_type="rgcn",
     seed=None,
     dropout=0,
     recurrent_dropout=0,
     num_tasks=1,
-    learning_rate_pre=0.005,
-    learning_rate_ft=0.01,
+    learning_rate = 0.01,
+    train_type = 'pre',
+    early_stop_rounds = False,
```

Comment:
Probably better set the default to `None`. As is, it suggests it's a boolean when in actuality it's supposed to be an `int`.
```suggestion
    early_stop_rounds = None,
```